blog/tech-talk-window-resize-behavior.md

```diff
@@ -0,0 +1,290 @@
+
+When we resize the window, what's supposed to happen is that in an atomic transaction (= at the exact same time) we repaint the viewport (= the area that's supposed to be visible on screen) and then update the clip rect to clip the surface to the new viewport size.
+
+After the resize, it should look like this:
+
+![Visualization with updated viewport and clip rect](/assets/img/blog/tech-talk-window-resize-behavior/visualization-2.png)
+
+And that's where we get to the first of our two bugs.
+
+## First bug
+
+Sometimes these operations can get out of sync. For example, the clip rect might get updated before the viewport is repainted. Then we get a result like this:
+
+![Visualization where the clip rect was updated before the viewport](/assets/img/blog/tech-talk-window-resize-behavior/visualization-3.png)
+
+We still show the old frame in the green viewport. But the clip rect has become larger and we show areas of the surface that we haven't repainted yet.
+
+On the first resize of a window, these areas would be black. On the second resize, those areas would be filled with old pixel values. They would show whatever we had previously painted in those areas.
+
+Similarly, in a certain edge case while making the window smaller, we would sometimes repaint the viewport before we would update the clip rect.
+
+![Visualization where the viewport was repainted before the clip rect was updated](/assets/img/blog/tech-talk-window-resize-behavior/visualization-4.png)
+
+Then parts of the clip rect would still show the previous frame because the new frame was smaller and we did not repaint any areas beyond the new viewport.
+
+Now why do these operations not happen in sync?
+
+We use two different Windows APIs here:
+
+- [`IDXGISwapChain1::Present1`](https://source.chromium.org/chromium/chromium/src/+/main:gpu/command_buffer/service/shared_image/dxgi_swap_chain_image_backing.cc;l=283;drc=92e802d8fa66bafb0cca9ed32143d6148d8e0411) — This shows the new pixels on screen / updates the new viewport.
+- [`IDCompositionDevice::Commit`](https://source.chromium.org/chromium/chromium/src/+/main:ui/gl/dc_layer_tree.cc;l=1034;drc=35d26d364efb57d0386b98312ba739f7f65ae97e) — This updates the clip rect.
+
+Here's what's important to understand: both functions return synchronously on the CPU. However, they schedule tasks that run asynchronously on the GPU at a later time. Windows and its services (such as [DWM](https://en.wikipedia.org/wiki/Desktop_Window_Manager)) decide when these tasks will run and in which order. So they take effect asynchronously, and not always within the same frame.
+
+Unfortunately, Windows provides no way for us to synchronize those operations. So I had to find other approaches to fix this.
+
+There were two options that I evaluated with the Chromium maintainers:
```

Comment:
Trimming out the discussion about different possible solutions and focusing on the one you implemented would be a good way to shorten the article and tighten it up.

Reply:
Thanks for the suggestion! I put it in because I thought it was actually interesting to see that these kinds of issues are solved in collaboration with Chromium maintainers. I'd prefer to leave it in.